### PR TITLE
Implement `FilterMap` iterator

### PR DESCRIPTION
**Pull request by MSE99**
**Please provide a brief description of the change.**

Adds `FilterMap` function and `FilterMapIter` type to iter package.

**Contribution checklist.**

_Replace the space in each box with "X" to check it off._

- [x] I have read and understood the CONTRIBUTING guidelines
- [x] My code is formatted (`make check`)
- [x] I have run tests (`make test`)
- [x] All commits in my PR conform to the commit hygiene section
- [x] I have added relevant tests
- [x] I have not added any dependencies

### CODE REVIEW

File: iter/filter_test.go [hunk 1]
```diff
@@ -52,3 +73,21 @@ func TestExclude(t *testing.T) {
 	assert.Equal(t, evens.Next().Unwrap(), 1)
 	assert.Equal(t, evens.Next().Unwrap(), 3)
 }
+
+func TestFilterMap(t *testing.T) {
```
Comment (BooleanCat) [inline iter/filter_test.go:77]: Please include a test for an empty underlying iterator

File: iter/filter.go [hunk 2]
```diff
@@ -36,3 +36,28 @@ func Exclude[T any](iter Iterator[T], fun func(T) bool) *FilterIter[T] {
 	inverse := func(t T) bool { return !fun(t) }
 	return &FilterIter[T]{iter, inverse}
 }
+
+type FilterMapIter[T any, U any] struct {
```
Comment (BooleanCat) [inline iter/filter.go:40]: Please include a compile time guarantee that the filter map iterator implements the iterator interface, see line 31.

File: iter/filter_test.go [hunk 3]
```diff
@@ -21,6 +22,26 @@ func ExampleFilter() {
...
 
+func ExampleFilterMap() {
+	selectAndTripleOdds := func(x int) option.Option[int] {
+		if x%2 == 0 {
+			return option.None[int]()
+		}
+
+		return option.Some(x * 3)
+	}
+	in := []int{1, 2, 3, 4, 5}
```
Comment (BooleanCat) [inline iter/filter_test.go:33]: Inline this so that the examples in the docs are as short as possible.

**Review by BooleanCat — CHANGES_REQUESTED** [review]
Thanks for the second contribution, there's a few changes I'd like for the tests.

File: iter/filter_test.go [hunk 4]
```diff
@@ -52,3 +69,28 @@ func TestExclude(t *testing.T) {
...
+
+		return option.Some(x * 2)
+	}
+
+	fltMap := iter.FilterMap[int](
+		iter.Lift([]int{1, 2, 3, 4, 5, 6}),
+		selectEvenAndDouble,
+	)
+	fltMapEmpty := iter.FilterMap[int](
+		iter.Lift([]int{}),
```
Comment (BooleanCat) [inline iter/filter_test.go:87]: iter.Exhausted will give an empty iterator.

File: iter/filter_test.go [hunk 5]
```diff
@@ -52,3 +69,28 @@ func TestExclude(t *testing.T) {
...
+		selectEvenAndDouble,
+	)
+
+	emptyResult := iter.Collect(fltMapEmpty)
+	result := iter.Collect(fltMap)
+
+	assert.SliceEqual(t, result, []int{4, 8, 12})
+	assert.SliceEqual(t, emptyResult, []int{})
```
Comment (BooleanCat) [inline iter/filter_test.go:95]: Use assert.Empty to test for empty slices.

File: iter/filter_test.go [hunk 6]
```diff
@@ -52,3 +69,28 @@ func TestExclude(t *testing.T) {
...
+		}
+
+		return option.Some(x * 2)
+	}
+
+	fltMap := iter.FilterMap[int](
+		iter.Lift([]int{1, 2, 3, 4, 5, 6}),
+		selectEvenAndDouble,
+	)
+	fltMapEmpty := iter.FilterMap[int](
```
Comment (BooleanCat) [inline iter/filter_test.go:86]: We should be testing only one behaviour in each test. Please separate this into two tests `TestFilterMap` and `TestFilterMapEmpty`.

**Review by BooleanCat — CHANGES_REQUESTED** [review]
In general we should validate one behaviour per test unless there's a good reason not to (such as slow tests).

A few more convenience changes too you may not have been aware of like for asserting empty slices and iter.Exhausted.

**Comment by BooleanCat** [comment]
Thanks for the second contribution.

I noticed with this PR and your previous PR that the Example tests were not functioning. I fix both of these myself and kept the git tree to preserve your contribution.

In case you weren't aware @MSE99, functions beginning with Example in test files serve two purposes: they are rendered into the reference docs and they are a functional test. The catch is that Go will ensure that stdout matches whatever follows `Output:` in comments at the end of the test. Since your tests didn't use "Output" nothing was being tested.

**Comment by MSE99** [comment]
@BooleanCat sorry for the broken example tests this is my first time using the example feature and my first time contributing to an open source go project, I'll do better next time thanks again
